developer/src/tike/main/UfrmMessages.pas

```diff
@@ -123,33 +125,145 @@ implementation
  - Message functions                                                           -
  -------------------------------------------------------------------------------}
 
+const // ABGR - text on white bg, Colors similar to Light+ VSCode color theme
+  Color_Filename = $96791A;
+  Color_Line     = $197975;
+  Color_Text     = $333333;
+  Color_Code     = $666666;
+
+  Color_Info     = $333333;
+  Color_Hint     = $A25015;
+  Color_Warn     = $156193;
+  Color_Error    = $3434C9;
+  Color_Fatal    = $3434C9;
+
+  Color_SuccessText  = $1C8816;
+  Color_InfoText     = $6B6B6B;
+  Color_FailureText  = $3434C9;
+
+  // Hard coded message numbers from infrastructureMessages.ts
+
+  NAMESPACE_Infrastructure = $5000;
+  INFO_FileBuiltSuccessfully = NAMESPACE_Infrastructure or $0006;
+  INFO_FileNotBuiltSuccessfully = NAMESPACE_Infrastructure or $0007;
+  INFO_ProjectBuiltSuccessfully = NAMESPACE_Infrastructure or $000B;
+  INFO_ProjectNotBuiltSuccessfully = NAMESPACE_Infrastructure or $000C;
+
+
 procedure TfrmMessages.Add(state: TProjectLogState; filename, msg: WideString; MsgCode, line: Integer);
+type
+  TSegment = record
+    text: string;
+    color: TColor;
+  end;
 var
   mi: TMessageItem;
   FColor: TColor;
+  FTextColor: TColor;
+  Segments: array of TSegment;
+
+  procedure AddText(const text: string; color: TColor);
+  var
+    s: TSegment;
+  begin
+    s.text := text;
+    s.color := color;
+    SetLength(Segments, Length(Segments)+1);
+    Segments[High(Segments)] := s;
+  end;
+
+  procedure AddLine;
+  var
+    t: string;
+    s: TSegment;
+    cr: TCharRange;
+    cf: TCharFormat;
+    gtle: TGetTextLengthEx;
+  begin
+    // TRichEdit's wrapper of the RichEdit control is very slow, so we instead
+    // use direct messages to the RichEdit control to emit formatted text
+    FillChar(cf, sizeof(TCharFormat), 0);
+    cf.cbSize := sizeof(TCharFormat);
+    cf.dwMask := CFM_COLOR;
+
+    gtle.flags := GTL_PRECISE or GTL_NUMCHARS;
+    gtle.codepage := 1200;
+    cr.cpMin := SendMessage(memoMessage.Handle, EM_GETTEXTLENGTHEX, NativeUint(@gtle), 0);
+    cr.cpMax := cr.cpMin;
+    SendMessage(memoMessage.Handle, EM_EXSETSEL, 0, NativeUint(@cr));
+
+    for s in Segments do
+      t := t + s.text;
+    memoMessage.SelText := t;
+
+    for s in Segments do
+    begin
+      cr.cpMin := cr.cpMax;
+      cr.cpMax := cr.cpMax + s.text.Length;
+      SendMessage(memoMessage.Handle, EM_EXSETSEL, 0, NativeUint(@cr));
+      cf.crTextColor := ColorToRGB(s.color);
+      SendMessage(memoMessage.Handle, EM_SETCHARFORMAT, SCF_SELECTION, NativeUint(@cf));
+    end;
+    cr.cpMin := cr.cpMax;
```

Comment:
What are cr.cpMin and cr.cpMax where they keep flipping? (ll 193, 201, 207)

Reply:
These are the start and end of the selection. Basically, we are moving along the text by the length of each segment, which is why we move cpMin to the previous cpMax, before incrementing the cpMax to the end of the next segment.